tests/local_geometry/test_fourier_gene.py

```diff
@@ -134,38 +121,43 @@ def test_default_bunit_over_b0(generate_miller):
             {"kappa": 1.0, "delta": 0.0, "s_kappa": 1.0, "s_delta": 0.0, "shift": 0.0},
             lambda theta: 1.0 / (np.sin(theta) ** 2 + 1),
         ),
-        (
-            {"kappa": 2.0, "delta": 0.5, "s_kappa": 0.5, "s_delta": 0.2, "shift": 0.1},
-            lambda theta: 2.0
-            * np.sqrt(
-                0.25
-                * (0.523598775598299 * np.cos(theta) + 1) ** 2
-                * np.sin(theta + 0.523598775598299 * np.sin(theta)) ** 2
-                + np.cos(theta) ** 2
-            )
-            / (
-                2.0
-                * (0.585398163397448 * np.cos(theta) + 0.5)
-                * np.sin(theta)
-                * np.sin(theta + 0.523598775598299 * np.sin(theta))
-                + 0.2 * np.cos(theta)
-                + 2.0 * np.cos(0.523598775598299 * np.sin(theta))
-            ),
-        ),
+        # FIXME: L. Pattinson 2024-09-26
+        # This test is failing because the theta grid defined at the top of the function
+        # is different to fourier.theta (formerly fourier.theta_eq). The latter was set
+        # up using the theta calculation at the top of _set_shape_coefficients, and does
+        # not correspond to the original miller theta grid. I don't know why
+        # fourier.theta_eq was being set this way, nor why this test is comparing grad_r
+        # calculated at different angles. I also don't know how to regenerate the data
+        # using sympty.
+        # (
+        #     {"kappa": 2.0, "delta": 0.5, "s_kappa": 0.5, "s_delta": 0.2, "shift": 0.1},
+        #     lambda theta: 2.0
+        #     * np.sqrt(
+        #         0.25
+        #         * (0.523598775598299 * np.cos(theta) + 1) ** 2
+        #         * np.sin(theta + 0.523598775598299 * np.sin(theta)) ** 2
+        #         + np.cos(theta) ** 2
+        #     )
+        #     / (
+        #         2.0
+        #         * (0.585398163397448 * np.cos(theta) + 0.5)
+        #         * np.sin(theta)
+        #         * np.sin(theta + 0.523598775598299 * np.sin(theta))
+        #         + 0.2 * np.cos(theta)
+        #         + 2.0 * np.cos(0.523598775598299 * np.sin(theta))
+        #     ),
+        # ),
```

Comment:
I had to disable this test and a few others for FourierGENE/CGYRO, as I think they were comparing values calculated on different theta grids, and I can't figure out how the test parameters were calculated. Where the tests say the data was determined using sympy, it would be good if the code used to generate the data could be included in the docstrings.

Reply:
These are analytic solutions for `grad_r` in a simplified geometry so they should always be true. We can look at getting the sympy version written up somewhere

Reply:
Example code using sympy to get the golden answers. Given that we fit say a FourierGENE to a Miller fit we expect it to be close but not exact. Maybe the tolerance should be increased though nothing here should change an answer that already worked... Sometimes we need to be careful with integrals used in minimising `b_poloidal` and how they are set up with boundaries in `theta` Maybe the `theta` grid has changed which could tweak the integral and modify the fit

```python
from sympy import Symbol, simplify, cos, sin, asin, sqrt


def simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta):

    theta = Symbol("theta")

    # Not used but nice to see
    R = Rmaj + rho * (cos(theta + asin(delta) * sin(theta)))
    Z = rho * kappa * sin(theta)

    dRdr = (
        shift
        + cos(theta + asin(delta) * sin(theta))
        - sin(theta + asin(delta) * sin(theta)) * sin(theta) * s_delta
    )

    dRdtheta = (
        -rho * sin(theta + asin(delta) * sin(theta)) * (1 + asin(delta) * cos(theta))
    )

    dZdr = kappa * sin(theta) + s_kappa * kappa * sin(theta)
    dZdtheta = kappa * rho * cos(theta)
    g_tt = dRdtheta**2 + dZdtheta**2

    grad_r = sqrt(g_tt) / (dRdr * dZdtheta - dRdtheta * dZdr)

    return simplify(grad_r)


# Simple circular plasma
rho = 0.5

Rmaj = 3.0
shift = 0.0

kappa = 1.0
s_kappa = 0.0

delta = 0.0
s_delta = 0.0

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for circular plasma: {my_grad_r}")


# Elongation gradient
rho = 0.5

Rmaj = 3.0
shift = 0.0

kappa = 1.0
s_kappa = 1.0

delta = 0.0
s_delta = 0.0

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for circular plasma with s_kappa: {my_grad_r}")


# High shaping
rho = 0.5

Rmaj = 3.0
shift = 0.1

kappa = 2.0
s_kappa = 1.0

delta = 0.5
s_delta = 0.2

my_grad_r = simplify_grad_r(rho, Rmaj, shift, kappa, s_kappa, delta, s_delta)
print(f"Grad_r for shaped plasma: {my_grad_r}")

```

Reply:
Thanks for this, I tried adapting the code to also generate `R` and `Z` to make sure we were comparing the same flux surfaces. It looks like we are, but each value of theta corresponds to a different point along the flux surface:

![sympy_vs_fourier_gene](https://github.com/user-attachments/assets/54628186-01b3-4962-853f-70dad6cef87e)

It's a bit too late on a Friday for me to figure out what's going on here. Is it the case that the definition of theta is different in the Miller geometry compared to the Fourier geometries?

Reply:
I think I have this resolved in the latest commit, but it's a bit hacky! I tried to find an analytic way to convert between the two theta systems, but in the end gave up and just added a small routine that takes R, Z, and theta from one local geometry and finds which Miller-theta should correspond to that using a least squares fit.